Add granular error list to alias action response

parkertimmins opened the pull request:
This commit changes the behavior when calling `POST _aliases` to remove an alias that does not exist. If the remove action has `must_exist==false` (or is not present), and a list of actions succeeds partially, the response will now contain a list of results for each input action.

For example on the request:
```
POST _aliases
{
  "actions": [
    {
      "remove": {
        "index": "new_index",
        "alias": "non_existing_alias",
        "must_exist": false
      }
    },
    {
      "add": {
        "index": "new_index",
        "alias": "other_alias"
      }
    }
  ]
}
```
if the first action fails and the second action succeeds, the response will be:
```
{
   "acknowledged": true,
   "errors": true,
   "action_results": [
      {
        "action”: "remove",
        "success:" false,
        "error": "alias_not_found_exception"
      },
      {
        "action”: "add",
        "success:" true,
      }
   ]
}
```


Additionally, this commit changes the error message returned when removing an non-existent alias with `must_exist==true`. It now throws a `alias_not_found_exception` so the behavior is the same as when `must_exist==false` and all actions fail.

This closes #94478 , though is different than the behavior requested in that ticket.

parkertimmins committed:
Add granular error list to alias action response

This commit does the following:
* When an alias action list succeeds partially a list of results for each
  action are returned. This will only occur is must_exist==false
* Use same exception for removal of missing aliases whether must_exist
  is true or false.

Closes #94478

parkertimmins commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/TransportIndicesAliasesAction.java:
```diff
@@ -106,15 +107,19 @@ protected void masterOperation(
         // Expand the indices names
         List<AliasActions> actions = request.aliasActions();
         List<AliasAction> finalActions = new ArrayList<>();
+        List<AliasActionResult> actionResults = new ArrayList<>();
```
I am a little uncomfortable with the change made in this file. Specifically, the use of the actionResult variable which is modified throughout the main loop over actions. Also, the fact the the action result is added to the final actionResults list at different places in the code. This is done because of the use of `continue`  in a few places. I think a better design would require a refactor of this loop to remove the use of `continue` statements. I was trying to avoid changing the code too much ... but kinda seems like I already passed that point.

parkertimmins replied:
:+1:  #107017

parkertimmins commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/IndicesAliasesResponse.java:
```diff
@@ -0,0 +1,136 @@
+            "aliases_not_found_exception"
+        );
+
+        private final AliasActions.Type actionType;
```
It might make sense to return the full action from the request rather than just the type. Thoughts?

joegallo replied:
Yeah, ++, nobody's going to remember what order they submitted what in, so even though you're returning things in the same order, you need to include enough information in the response for them to understand what failed.

parkertimmins replied:
Does **type** (add/remove), **indices**, and **aliases** seem like enough information? After some digging, I've found returning the original request is problematic because it gets modified. I can add other fields that are important, but it seems better to return fewer fields, as adding fields is non-breaking, but removing/update fields is a breaking change.

parkertimmins committed:
Add response body description to aliases API docs

elasticsearchmachine commented:
Hi @parkertimmins, I've created a changelog YAML for you.

parkertimmins committed:
Update docs/changelog/106514.yaml

parkertimmins committed:
Add new transport version and fix broken tests

parkertimmins committed:
Formatting - need to call spotlessApply

parkertimmins committed:
Skip AliasActionResult yaml rest tests before 8.14

parkertimmins committed:
Merge branch 'main' into alias-remove-must-exist

masseyke commented on docs/reference/indices/aliases.asciidoc:
```diff
@@ -168,3 +174,38 @@ stream aliases don't support this parameter.
+
+`success`::
+(boolean)
+If true, the requested action was successfully applied at least once.
```
At least once? We potentially run the actions they ask for repeatedly?

parkertimmins replied:
Kind of. Since the remove action can have wildcards, it can apply to multiple aliases. It's treated as a success if one or more aliases match. Maybe it would be clearer to say: `the requested action was successfully applied to one or more aliases`

parkertimmins committed:
Fix broken compilation after merge from main

masseyke commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/TransportIndicesAliasesAction.java:
```diff
@@ -222,25 +238,33 @@ protected void masterOperation(
                         throw new IllegalArgumentException("Unsupported action [" + action.actionType() + "]");
                 }
             }
+            actionResults.add(actionResult);
```
It took me a few minutes to convince myself that `actionResult` can never be null here, so I thought I'd write down why in order to possibly save someone else some time.
It can't because `concreteIndices` is never empty because we would have already had an exception by this point if it had been because IndicesAliasesRequest's indicesOptions always sets `allowNoIndices` to `false`.

parkertimmins replied:
Thanks for reasoning this out. This code is pretty confusing, and would be easy to introduce a bug (assuming there's not one already!).
I just refactored this loop in a way that I think is easier to believe is bug free.

parkertimmins committed:
Pass on actual response rather than default value

parkertimmins committed:
Clarify confusing phrase in docs

parkertimmins committed:
Refactor main alias action result loop

The current logic in this main loop is really hard to track. It is not
obvious that the action results are guaranteed to not be null. It would
also be easy to introduce a change that does make them null. Hopefully,
this changes makes to more clear that the alias action results lists is
built safely.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

masseyke commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/TransportIndicesAliasesAction.java:
```diff
@@ -161,18 +165,21 @@ protected void masterOperation(
                                 finalActions.add(new AddDataStreamAlias(alias, dataStreamName, action.writeIndex(), action.filter()));
                             }
                         }
+                        actionResults.add(AliasActionResult.build(action, numAliasesRemoved));
```
This doesn't hurt anything, but it threw me off for a second that we're passing `numAliasesRemoved` when building an `add` action result. AliasActionResult.build correctly ignores it.

parkertimmins replied:
Yeah, that does look weird. I think it's worth using separate `build` and `buildRemoveAlias` builder functions, so it's a bit clearer.

masseyke commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/TransportIndicesAliasesAction.java:
```diff
@@ -106,15 +107,18 @@ protected void masterOperation(
         // Resolve all the AliasActions into AliasAction instances and gather all the aliases
         Set<String> aliases = new HashSet<>();
         for (AliasActions action : actions) {
+            int numAliasesRemoved = 0;
```
It's out of scope for this PR, but I really think we ought to refactor this method in another PR. For a block of code that does something so simple, It is incredibly confusing to keep track of what is going on in here.

parkertimmins committed:
Add info from request to alias action response

Ideally we would return the AliasActions request directly to the users.
But the AliasActions object contains logic which we probably don't want
to expose in an api. Also,  index patterns are
resolved in the security layer (if its on), meaning the indices field of the
request does not contain the pattern provided by the user. To solve this
we store the original indices and do not update this field when the
security update the indices field

parkertimmins commented on server/src/main/java/org/elasticsearch/action/admin/indices/alias/IndicesAliasesRequest.java:
```diff
@@ -301,6 +317,9 @@ public AliasActions index(String index) {
             this.indices = new String[] { index };
+            if (this.originalIndices == null) {
+                this.originalIndices = this.indices;
+            }
```
There is an `originalAliases` field which is created for the same purpose: to keep the aliases from the request before they are resolved by the security layer. Unfortunately, despite being analogous, `orginalIndices` and `orginalAliases` are handled slightly differently. This is because the `indices` is modified in the security layer by calling the `indices()` function a second time which updates the field; thus we must check if `originalIndices` is null to avoid setting it a second time. This differs from how aliases are set. They are set the first time (along with `originalAliases`) in the `aliases()` function, then only the `aliases` field (and not `originalAliases`) is updated in the `replaceAliases()` function.

Anyway, I'd prefer that this code behaved the same as the similar code for aliases, but I think making such a change would be a too far reaching change.

parkertimmins replied:
Yikes, yeah if the request is reused this will not work correctly. I can't think of any way to store the unresolved value which will work with reuse of the request. (Well we could change the security layer's call from `indices()` to a new function `replaceIndices()` as is done for aliases, but that seems well out of scope).

Using security-substituted index names isn't horrible. Unfortunately, the index and alias names are not resolved in the request when it enters TransportIndicesAliasesAction if security is disabled. This class calls out to the resolver. To keep security and non-security behavior identical, some tracking of index/alias resolved names would need to be added to the already complicated main loop.

Do you think it would be crazy to just not include any information about the action in the response? And require the user correlate the result list with the request list? Joe correctly pointed out the that the user cannot be expected to still have the request around ... but no information might be better than inconsistent information.

dakrone replied:
I think using the `originalIndices` behavior as it is in this PR is more confusing and possibly incorrect. For now I think it'd be better to use the security-substituted index name, and perhaps track the originals differently the same work as #107017, what do you think?

parkertimmins committed:
Update docs with action info added in last commmit

parkertimmins committed:
Merge branch 'main' into alias-remove-must-exist

parkertimmins committed:
Merge branch 'main' into alias-remove-must-exist

masseyke reviewed (approved):
Looks good to me. I'm curious if @joegallo or @dakrone have any other comments on it.

joegallo commented:
I will take a look on Monday, if you're willing to wait until then, @parkertimmins.

dakrone commented:
I briefly looked at this with some manual testing, but it appears to break the `?error_trace` parameter. Prior to this change, I can do:

```
POST _aliases?error_trace
{
  "actions": [
    {
      "remove": {
        "index": "index1",
        "alias": "logs-non-existing",
        "must_exist": true
      }
    },
    {
      "add": {
        "index": "index2",
        "alias": "logs-non-existing"
      }
    }
  ]
}
```

And get back the full stack trace, but after this change, even adding the `?error_trace` still only gives me an error that looks like:

```
{
  "acknowledged" : true,
  "errors" : true,
  "action_results" : [
    {
      "action" : {
        "type" : "remove",
        "indices" : [
          "index1"
        ],
        "aliases" : [
          "logs-non-existing"
        ]
      },
      "success" : false,
      "error" : "aliases_not_found_exception"
    },
    {
      "action" : {
        "type" : "add",
        "indices" : [
          "index2"
        ],
        "aliases" : [
          "logs-non-existing"
        ]
      },
      "success" : true
    }
  ]
}
```

parkertimmins commented:
Discussed with @dakrone and decided this api should behave more like the bulk api. 

When `error_trace` is used with the bulk api, the stack trace is not returned for each error result within a 200 response. `error_trace` seems to only return a full stack trace when the response is a 400. The alias action api behaves the same, so is fine wrt `error_trace`.

On the other hand, a couple things should be changed to match the bulk api. The `error` string field, should be replaced with an actual exception object. This will be the same exception type as is thrown when `must_exist==true` or all actions fail. Also, the boolean `success` field should be a http `status` field. 

For example this would the response to an alias action list with a failure and a success: 
```
{
  "acknowledged": true,
  "errors": true,
  "action_results": [
    {
      "action": {
        "type": "remove",
        "indices": [
          "index1"
        ],
        "aliases": [
          "logs-non-existing"
        ]
      },
      "status": 404,
      "error": {
        "type": "aliases_not_found_exception",
        "reason": "aliases [logs-non-existing] missing",
        "resource.type": "aliases",
        "resource.id": "logs-non-existing"
      }
    },
    {
      "action": {
        "type": "add",
        "indices": [
          "index2"
        ],
        "aliases": [
          "logs-non-existing"
        ]
      },
      "status": 200
    }
  ]
}
```

parkertimmins committed:
Change success+error msg to status+error object

parkertimmins committed:
Fix yaml rest tests

parkertimmins committed:
update docs with status + error obj

parkertimmins committed:
Merge branch 'main' into alias-remove-must-exist

parkertimmins committed:
Merge branch 'main' into alias-remove-must-exist

dakrone reviewed (changes requested):
I left just a few comments, but this is very close, thanks for iterating on the output format with me!

parkertimmins committed:
Revert originalIndices logic to use indices

If a request is reused, the originalIndices field will
be incorrect. Instead, use indices field. It will contain
the indices resolved by the security layer.

parkertimmins committed:
Add resolved index names to alias action result

Resolved index names are already provided when security layer
is used. This adds resolved indices to alias action result
in all cases.

parkertimmins committed:
Merge branch 'main' into alias-remove-must-exist

parkertimmins committed:
Fix tests to expect index name resolution

parkertimmins committed:
Add serialization tests for IndicesAliasesResponse

parkertimmins committed:
Test serialization Ack <-> IndicesAliasesResponse

parkertimmins committed:
add javadoc

parkertimmins committed:
Asserts on IndiceAliasesResponse construction

IndicesAliasesResponse could be instantiated with an empty error list
and set the errors field to true. Add some validation to make sure is
not instantiated this way.

parkertimmins committed:
Add exception class to equality/hashcode

Rather than checking exception equality exactly, just
make sure exception class and message are equal

parkertimmins committed:
spotlessApply

parkertimmins committed:
Remove incorrect assert

There should be one AliasActionResult for every AliasActions, not
for every AliasAction. The former is the requested action on the
 unresolved alias pattern, which may map to multiple of the latter
 each of which is for a single specific alias

parkertimmins committed:
Merge branch 'main' into alias-remove-must-exist

dakrone reviewed (approved):
LGTM, thanks for iterating on this @parkertimmins!

parkertimmins committed:
Merge branch 'main' into alias-remove-must-exist
